### PR TITLE
EndpointSlice 1.22 Docs Update (KEP 752)

### PR DESCRIPTION
**Pull request by swetharepakula**
This PR includes the doc updates required since EndpointSlice is going GA in 1.22.

* KEP: EndpointSlice Enhancement: [EndpointSlice KEP](https://github.com/kubernetes/enhancements/blob/8413469b89851d094eb22813a06594bd4a6d36a4/keps/sig-network/0752-endpointslices/README.md)
* Enhancement Issue: https://github.com/kubernetes/enhancements/issues/752
* Relevant Kubernetes PRs:
  * kubernetes/kubernetes#103306
  * kubernetes/kubernetes#103451
  * kubernetes/kubernetes#103520

### CODE REVIEW

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 1]
```diff
@@ -406,6 +406,9 @@ different Kubernetes components.
 | `WatchBookmark` | `false` | Alpha | 1.15 | 1.15 |
 | `WatchBookmark` | `true` | Beta | 1.16 | 1.16 |
 | `WatchBookmark` | `true` | GA | 1.17 | - |
+| `WindowsEndpointSliceProxying` | `false` | Alpha | 1.19 | 1.20 |
+| `WindowsEndpointSliceProxying` | `true` | Beta | 1.21 | 1.21 |
+| `WindowsEndpointSliceProxying` | `true` | GA | 1.22| - |
```
Comment (aojea) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:411]: what about `EndpointSliceProxying`

also, it seems that the code is still in beta
```
pkg/features/kube_features.go:  EndpointSliceProxying featuregate.Feature = "EndpointSliceProxying"
pkg/features/kube_features.go:  WindowsEndpointSliceProxying featuregate.Feature = "WindowsEndpointSliceProxying"
pkg/features/kube_features.go:  EndpointSliceProxying:                          {Default: true, PreRelease: featuregate.Beta},
pkg/features/kube_features.go:  WindowsEndpointSliceProxying:                   {Default: true, PreRelease: featuregate.Beta},
```

**Comment by PI-Victor** [comment]
thank you for opening this, please remember to hold until the enhancement is no longer labelled 'At Risk'.

/sig network
/assign @ritpanjw

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 1 above)
Comment (PI-Victor) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:411]: one thing to note, if it is GA, it needs to be removed from the first table as well

**Comment by sftim** [comment]
https://github.com/kubernetes/kubernetes/pull/103596 suggests this PR might also need to cover `EndpointSliceTerminatingCondition`.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 1 above)
Comment (swetharepakula) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:411]: This draft is still very much a WIP. I just opened the PR with a small change for tracking purposes.

**Comment by swetharepakula** [comment]
The enhancement is no longer at risk. All relevant PRs were merged yesterday.

This PR is to cover the changes related to https://github.com/kubernetes/enhancements/issues/752

kubernetes/kubernetes#103596 is for a different KEP. I believe `EndpointSliceTerminatingCondition` is for https://github.com/kubernetes/enhancements/issues/1672 and will be covered by a different docs PR

**Comment by ritpanjw** [comment]
Hello @swetharepakula :wave: please take a look at [Documenting for a release - PR Ready for Review to get your PR ready](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) for review before **Tuesday 20th of July, 11:59 PM PDT**.  In comments, you mentioned it is WIP, hence sending reminder. Thank you!

**Comment by PI-Victor** [comment]
> Hello @swetharepakula 👋 please take a look at [Documenting for a release - PR Ready for Review to get your PR ready](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) for review before **Tuesday 20th of July, 11:59 PM PDT**. In comments, you mentioned it is WIP, hence sending reminder. Thank you!

Hi @swetharepakula friendly reminder that today is the deadline for PRs ready for review, please make sure this PR is ready for review by the end of the day, thank you!

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 1 above)
Comment (swetharepakula) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:445]: Gates have graduated in the code. And the gates have been removed from the first table.

**Comment by swetharepakula** [comment]
Thanks for the reminder @ritpanjw & @PI-Victor! The PR is now ready for review

**Comment by robscott** [comment]
This PR LGTM from technical side, one question for docs reviewers like @sftim and @kbhawkey: Do we ever allow docs to reference previous versions? In this case it might be helpful to note that the `warning` annotation value would be set in 1.21 if users want to know if any of their Endpoints will be truncated when they upgrade to 1.22.

**Comment by sftim** [comment]
> Do we ever allow docs to reference previous versions?

Usually, the way we'd help readers is by advising them to check they're viewing the docs for _their_ version of Kubernetes.
We can backport advice to older docs revisions if that makes sense to do so.

**Comment by aojea** [comment]
LGTM

**Comment by robscott** [comment]
>Usually, the way we'd help readers is by advising them to check they're viewing the docs for their version of Kubernetes.
We can backport advice to older docs revisions if that makes sense to do so.

Thanks for the quick response @sftim! That approach makes sense to me. I think the docs for 1.21 are already sufficient here.

File: content/en/docs/concepts/services-networking/service.md [hunk 2]
```diff
@@ -188,9 +188,10 @@ selectors and uses DNS names instead. For more information, see the
 [ExternalName](#externalname) section later in this document.
 
 ### Over Capacity Endpoints
-If an Endpoints resource has more than 1000 endpoints then a Kubernetes v1.21 (or later)
-cluster annotates that Endpoints with `endpoints.kubernetes.io/over-capacity: warning`.
-This annotation indicates that the affected Endpoints object is over capacity.
+If an Endpoints resource has more than 1000 endpoints then a Kubernetes v1.22 (or later)
+cluster annotates that Endpoints with `endpoints.kubernetes.io/over-capacity: truncated`.
```
Comment (kbhawkey) [inline content/en/docs/concepts/services-networking/service.md:192]: nit: Can you be more specific? What annotates the Endpoints resource?

File: content/en/docs/reference/labels-annotations-taints.md [hunk 3]
```diff
@@ -262,11 +262,11 @@ The value of the annotation is the container name that is default for this Pod.
 
 ## endpoints.kubernetes.io/over-capacity
 
-Example: `endpoints.kubernetes.io/over-capacity:warning`
+Example: `endpoints.kubernetes.io/over-capacity:truncated`
 
 Used on: Endpoints
 
-In Kubernetes clusters v1.21 (or later), the Endpoints controller adds this annotation to an Endpoints resource if it has more than 1000 endpoints. The annotation indicates that the Endpoints resource is over capacity.
+In Kubernetes clusters v1.22 (or later), the Endpoints controller adds this annotation to an Endpoints resource if it has more than 1000 endpoints. The annotation indicates that the Endpoints resource was over capacity and the number of endpoints has been truncated to 1000.
```
Comment (kbhawkey) [inline content/en/docs/reference/labels-annotations-taints.md:269]: nit: "is" seems okay?

File: content/en/docs/concepts/services-networking/service.md [hunk 4]
```diff
@@ -188,9 +188,10 @@ selectors and uses DNS names instead. For more information, see the
 [ExternalName](#externalname) section later in this document.
 
 ### Over Capacity Endpoints
-If an Endpoints resource has more than 1000 endpoints then a Kubernetes v1.21 (or later)
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/service.md:191]: We should maybe edit the v1.21 docs (either now or post release) to remove the “or later” part from the v1.21 docs. WDYT @swetharepakula ?

Comment (swetharepakula) [reply content/en/docs/concepts/services-networking/service.md:191]: Oh that is a good point! I can send in a followup pr for that. Would branch would that pr need to be against?

File: content/en/docs/reference/labels-annotations-taints.md (hunk 3 above)
Comment (swetharepakula) [reply content/en/docs/reference/labels-annotations-taints.md]: Done.

File: content/en/docs/concepts/services-networking/service.md (hunk 2 above)
Comment (swetharepakula) [reply content/en/docs/concepts/services-networking/service.md:192]: Added a small addition stating that the endpoint controller truncates the resource. Please let me know if you think we should expand more. I am not sure how detailed I should be here.

File: content/en/docs/concepts/services-networking/service.md (hunk 4 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md:191]: At the moment, that is the _main_ branch: commits there are continuously published to https://kubernetes.io/

**Comment by sftim** [comment]
Next step is a tech review from SIG Network
@kubernetes/sig-network-pr-reviews please check this one out - it should be straightforward to verify.

File: content/en/docs/concepts/services-networking/service.md (hunk 2 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/service.md:192]: “the control plane has truncated” would be fine.

**Review by sftim — COMMENTED** [review]
/lgtm

File: content/en/docs/concepts/services-networking/service.md (hunk 4 above)
Comment (swetharepakula) [reply content/en/docs/concepts/services-networking/service.md:191]: Corresponding PR: https://github.com/kubernetes/website/pull/29122

**Comment by sftim** [comment]
/approve
/hold
Release docs folks, feel free to unhold

**Comment by PI-Victor** [comment]
@swetharepakula can you please rebase, a fix that fixes the feature gates merged and cause this to conflict.

**Comment by swetharepakula** [comment]
PR has been rebased.

Is there anything left for this PR? Or do we still need the hold?

**Comment by PI-Victor** [comment]
a tech LGTM from @kubernetes/sig-network-pr-reviews is all that is needed, when this is done we can remove the hold.

**Comment by sftim** [comment]
We have the tech review I think: https://github.com/kubernetes/website/pull/28745#issuecomment-883752047
/hold cancel

**Pull request merged**